File: C++/Recursion/quick_sort.cpp
#include <bits/stdc++.h>

using namespace std;

void print(vector<int> arr, int low, int high)
{
    if (low == high)
        return;

    cout << arr[low] << " ";
    print(arr, low + 1, high);
}

int getPartition(vector<int> &arr, int low, int high)
{

    int pVot = arr[low];
    int i = low;
    int j = high;

    while (i < j)
    {
        while (arr[i] <= pVot && i <= high - 1)
        {
            i++;
        }

        while (arr[j] > pVot && j >= low + 1)
        {
            j--;
        }

        if (i < j)
            swap(arr[i], arr[j]);
    }

    swap(arr[low], arr[j]);
    return j;
}

void quickSort(vector<int> &arr, int low, int high)
{
    if (low >= high)
        return;

    int partitionIndex = getPartition(arr, low, high);
    quickSort(arr, low, partitionIndex - 1);
    quickSort(arr, partitionIndex + 1, high);

}

int main()
{

    vector<int> arr = {5, 7, 3, 2, 9, 6};

    print(arr, 0, arr.size());
    cout << endl;

    quickSort(arr, 0, arr.size() - 1);

    print(arr, 0, arr.size());

    return 0;
}

File: Python/Sorting/quick_sort.py
def partition(a, low, high):
    pivot = a[low]
    i = low
    j = high

    while i < j:
        while i <= high - 1 and a[i] <= pivot:
            i += 1
        while j >= low + 1 and a[j] > pivot:
            j -= 1

        if i < j:
            a[i], a[j] = a[j], a[i]

    a[low], a[j] = a[j], a[low]
    return j


def quickSort(a, low, high):
    if low < high:
        j = partition(a, low, high)
        quickSort(a, low, j - 1)
        quickSort(a, j + 1, high)


if __name__ == '__main__':
    a = [10, 16, 8, 12, 15, 6, 3, 9, 5]

    print(a)

    quickSort(a, 0, len(a) - 1)

    print(a)
